drivers/serial/uart_opentitan.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023 by Rivos Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/drivers/uart.h>
+#include <soc.h>
+#include <zephyr/irq.h>
+
+/* Register offsets within the UART device register space. */
+#define UART_INTR_STATE_REG_OFFSET 0x0
+#define UART_INTR_ENABLE_REG_OFFSET 0x4
+#define UART_CTRL_REG_OFFSET 0x10
+#define UART_STATUS_REG_OFFSET 0x14
+#define UART_RDATA_REG_OFFSET 0x18
+#define UART_WDATA_REG_OFFSET 0x1c
+#define UART_FIFO_CTRL_REG_OFFSET 0x20
+#define UART_OVRD_REG_OFFSET 0x28
+#define UART_TIMEOUT_CTRL_REG_OFFSET 0x30
+
+/* Control register bits. */
+#define UART_CTRL_TX_BIT 0
+#define UART_CTRL_RX_BIT 1
+#define UART_CTRL_NCO_OFFSET 16
+
+/* FIFO control register bits. */
+#define UART_FIFO_CTRL_RXRST_BIT 0
+#define UART_FIFO_CTRL_TXRST_BIT 1
+
+/* Status register bits. */
+#define UART_STATUS_TXFULL_BIT 0
+#define UART_STATUS_RXEMPTY_BIT 5
+
+#define DT_DRV_COMPAT lowrisc_opentitan_uart
+
+#define DEV_CFG(dev) \
+	((const struct uart_opentitan_config *const)(dev)->config)
+
+struct uart_opentitan_config {
+	mem_addr_t base;
+	uint32_t nco_reg;
+};
+
+static int uart_opentitan_init(const struct device *dev)
+{
+	const struct uart_opentitan_config *cfg = DEV_CFG(dev);
+	/* Reset settings. */
+	sys_write32(0u, cfg->base + UART_CTRL_REG_OFFSET);
+	/* Clear FIFOs. */
+	sys_write32((1 << UART_FIFO_CTRL_RXRST_BIT) |
+		    (1 << UART_FIFO_CTRL_TXRST_BIT),
+		    cfg->base + UART_FIFO_CTRL_REG_OFFSET);
+	/* Clear other states. */
+	sys_write32(0u, cfg->base + UART_OVRD_REG_OFFSET);
+	sys_write32(0u, cfg->base + UART_TIMEOUT_CTRL_REG_OFFSET);
+	/* Disable interrupts. */
+	sys_write32(0u, cfg->base + UART_INTR_ENABLE_REG_OFFSET);
+	/* Clear interrupts. */
+	sys_write32(0xffffffffu, cfg->base + UART_INTR_STATE_REG_OFFSET);
+	/* Set baud and enable TX and RX. */
+	sys_write32((1 << UART_CTRL_TX_BIT) |
+		(1 << UART_CTRL_RX_BIT) |
+		(cfg->nco_reg << UART_CTRL_NCO_OFFSET),
+		cfg->base + UART_CTRL_REG_OFFSET);
+	return 0;
+}
+
+static int uart_opentitan_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct uart_opentitan_config *cfg = DEV_CFG(dev);
+
+	if (sys_read32(cfg->base + UART_STATUS_REG_OFFSET) &
+	    (1 << UART_STATUS_RXEMPTY_BIT)) {
+	    /* Empty RX FIFO */
+		return -1;
+	}
+	*c = sys_read32(cfg->base + UART_RDATA_REG_OFFSET);
+	return 0;
+}
+
+static void uart_opentitan_poll_out(const struct device *dev, unsigned char c)
+{
+	const struct uart_opentitan_config *cfg = DEV_CFG(dev);
+
+	/* Wait for space in the TX FIFO */
+	while (sys_read32(cfg->base + UART_STATUS_REG_OFFSET) &
+		(1<<UART_STATUS_TXFULL_BIT)) {
+	}
```

Comment:
nit: use `BIT()` macro (everywhere) and change to:
```c
while (whatever) {
    ;
}
```

Reply:
Done.